### ISSUE TITLE
templates/jqueryPlugin.js doesn't seem to conform to CommonJS spec

### ISSUE THREAD

**Issue by malthejorgensen** [umdjs/umd#87]
Not sure if I'm looking at the right [CommonJS spec](http://wiki.commonjs.org/wiki/Modules/1.1.1), but in the current jQuery plugin template ([jqueryPlugin.js, line 7](https://github.com/umdjs/umd/blob/master/templates/jqueryPlugin.js#L7)) `module.exports` is used instead of the CommonJS standard `exports`.

You can compare this with [commonjsStrict, line 23](https://github.com/umdjs/umd/blob/master/templates/commonjsStrict.js#L23), which does check and use the `exports`-object – not `module.exports`.

I think `jqueryPlugin.js` should be changed to use the `exports`-object.

_Note: Both files were checked at commit 95563fd6b46f06bda0af143ff67292e7f6ede6b7 (latest master as of 23rd Dec 2016)_

**Comment by kottenator** [comment]
@malthejorgensen - no, [it's fine](https://nodejs.org/api/modules.html#modules_module_exports), there are both `module.exports` and `exports` exist in CommonJS. The difference is:

* `module.exports = smth` provides default exported object for the whole module. It will be used as `var smth = require('my-module')`
* `exports.smth = smth` is for _multiple exports_ from the same module (e.g.): `var smth = require('my-module').smth`.

## Real problem with UMD for jQuery plugin and CommonJS

There is different problem with CommonJS and current `jqueryPlugin.js` implementation:

```js
if (typeof module === 'object' && module.exports) {
  // Node/CommonJS
  module.exports = function(root, jQuery) {
    if (jQuery === undefined) {
      // require('jQuery') returns a factory that requires window to
      // build a jQuery instance, we normalize how we use modules
      // that require this pattern but the window provided is a noop
      // if it's defined (how jquery works)
      if (typeof window !== 'undefined') {
        jQuery = require('jquery');
      }
      else {
        jQuery = require('jquery')(root);
      }
    }
    factory(jQuery);
    return jQuery;
  };
}
```

So when you do `var plugin = require('my-plugin')` - it's actually a _factory function_ which you need to _call_ to initialize the plugin: 

```js
plugin(); // or ...
plugin(window); // or ...
plugin(window, $); // or ...
plugin(null, $);
```

Looks complicated. I don't understand - why not to do simply:

```js
if (typeof module === 'object' && module.exports) {
  // Node/CommonJS
  var jQuery = require('jquery');
  factory(jQuery);
  module.exports = jQuery;
}
```

But the problem described above is not a bug - just do `require('my-plugin')()` instead of `require('my-plugin')`.

**Comment by malthejorgensen** [comment]
Hi @kottenator 

By that logic CommonJS and Node.js-modules are the same – I don't think that is true – have you read the [CommonJS spec](http://wiki.commonjs.org/wiki/Modules/1.1.1) – it doesn't mention `module.exports`?

I think the reality is that Node.js-modules added `module.exports` for convenience, and uses that as the "single source of truth", and then has the `exports` variable reference `module.exports`. This makes any CommonJS module work in Node.js – but not the other way around. A Node.js module could use _only_ `module.exports` but that's not compatible with the CommonJS spec, which only takes into account the `exports` variable.

Maybe the "CommonJS" label should be removed (and replaced with "Node.js-compatible") from the UMD templates?

**Comment by kottenator** [comment]
Agree, `module.exports` is not by CommonJS spec, you're right. But that's how it works in Node.js and I don't know other way to _default_ export in ES5.

**Comment by mike3run** [comment]
Hey guys I faced that problem yesterday what I did is this:

```js
;(function (factory) {
  if (typeof define === 'function' && define.amd) {
    define(['jquery'], factory)
  } else if (typeof exports !== 'undefined') {
    module.exports = factory(require('jquery'))
  } else {
    factory(jQuery)
  }
}(function ($) {
  // All the normal plugin here
}))
```

Insert your normal `;` if you'd like I removed them since I use `standard` as my lint tool

**Comment by alexweissman** [comment]
All, I've run into an issue with this template in Select2 (see https://github.com/select2/select2/commit/45a877345482956021161203ac789c25f40a7d5e#commitcomment-24593031).

Any guidance would be appreciated, as Select2 is a fairly large community and this issue could potentially be affecting a lot of users.
